### ISSUE TITLE
R Markdown websites share same environment when using `render_site()`?

### ISSUE THREAD

**Issue by andrewheiss** [rstudio/rmarkdown#1326]
This may be a bug, but it also might be the intended behavior for R Markdown. This comes from [an issue I posted on StackOverflow](https://stackoverflow.com/questions/49930372/use-new-environments-for-each-knitted-file-in-an-r-markdown-website). The reason I think it might be a bug is that when `blogdown` and `bookdown` compile a bunch of Rmd files, each one gets its own new environment, while `rmarkdown::render_site()` seems(?) to use one shared environment for all the Rmd files when knitting, which causes unexpected namespace issues.

For example, using this minimum working example with the following 5 files:

**testing.Rproj**

    Version: 1.0

    RestoreWorkspace: Default
    SaveWorkspace: Default
    AlwaysSaveHistory: Default

    EnableCodeIndexing: Yes
    UseSpacesForTab: Yes
    NumSpacesForTab: 2
    Encoding: UTF-8

    RnwWeave: Sweave
    LaTeX: pdfLaTeX

    AutoAppendNewline: Yes

    BuildType: Website

**\_site.yml**

    name: "testing"
    navbar:
      title: "Testing"
      left:
        - text: "Home"
          href: index.html
        - text: "Step 1"
          href: 01_something.html
        - text: "Step 2"
          href: 02_something-else.html

**index.Rmd**

    ---
    title: "Testing"
    ---

    Hello, Website!

**01_something.Rmd**

    ---
    title: "Step 1"
    ---

    Write some data just for fun.

    ```{r}
    library(tidyverse)
    library(here)

    write_csv(mtcars, file.path(here(), "cars.csv"))
    ```

**02_something-else.Rmd**

    ---
    title: "Step 2"
    ---

    This breaks because `lubridate::here()` and `here::here()` conflict, but *only* when rendering the whole site.

    ```{r}
    library(tidyverse)
    library(lubridate)
    library(here)

    # Do something with lubridate
    my_date <- ymd("2018-04-19")

    # Try to use here() and it breaks when rendering the whole site
    # It works just fine when knitting this file on its own, though, since here is loaded after lubridate
    cars <- read_csv(file.path(here(), "cars.csv"))
    ```

Both `here` and `lubridate` have a `here()` function, and because I want to use `lubridate`'s throughout the script in `02_something-else.Rmd`, I run `library(here)` after `library(lubridate)`. Running `02_something-else.Rmd` interactively or knitting it by itself works just fine—package loading happens in the right order and everything's great.

However, when building the site with `rmarkdown::render_site()` (from the console, from the "Build" button in RStudio, or from the terminal with `Rscript -e "rmarkdown::render_site())"`) there's an error when R gets to `02_something-else.Rmd`:

    Error: '2018-04-19 14:53:59/cars.csv' does not exist in current working directory

Instead of using `here::here()`, R is using `lubridate::here()` and inserting the current date and time, since `library(here)` was originally loaded in `01_something.Rmd` and that environment seems to still be loaded when R gets to `02_something-else.Rmd`.

According to [the documentation for `rmarkdown::render_site()`](https://www.rdocumentation.org/packages/rmarkdown/versions/1.9/topics/render_site), you can use the `envir = new.env()` argument to ensure that the site rendering uses a new environment, but that doesn't fix the problem. It seems that guarantees a new environment for the overall site building process, but not for the individual files.

Is there a way to ensure that *each individual file* in an R Markdown website gets its own new environment when knitted?

For now I've fixed the issue by prefixing all my lubridate functions with `lubridate::`, but this can cause more complicated namespace issues later on. For instance, `dplyr::filter()` and `MASS::filter()` commonly conflict with each other. If I use MASS in one file in the site, it would mess up code in all the files that come after, but I wouldn't realize it until I build the whole website.

`blogdown` and `bookdown` don't have this issue—it seems to be limited to just R Markdown websites. Perhaps this is by design, but I don't know.

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.4 (2018-03-15)
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  tz       America/Denver              
#>  date     2018-04-19
#> Packages -----------------------------------------------------------------
#>  package    * version    date       source                            
#>  assertthat   0.2.0      2017-04-11 CRAN (R 3.4.0)                    
#>  backports    1.1.2      2017-12-13 CRAN (R 3.4.3)                    
#>  base       * 3.4.4      2018-03-15 local                             
#>  bindr        0.1.1      2018-03-13 CRAN (R 3.4.4)                    
#>  bindrcpp     0.2        2017-06-17 CRAN (R 3.4.0)                    
#>  broom        0.4.3      2017-11-20 CRAN (R 3.4.2)                    
#>  cellranger   1.1.0      2016-07-27 CRAN (R 3.4.0)                    
#>  cli          1.0.0      2017-11-05 CRAN (R 3.4.2)                    
#>  colorspace   1.3-2      2016-12-14 CRAN (R 3.4.0)                    
#>  compiler     3.4.4      2018-03-15 local                             
#>  crayon       1.3.4      2017-09-16 CRAN (R 3.4.1)                    
#>  datasets   * 3.4.4      2018-03-15 local                             
#>  devtools     1.13.4     2017-11-09 CRAN (R 3.4.2)                    
#>  digest       0.6.15     2018-01-28 cran (@0.6.15)                    
#>  dplyr      * 0.7.4      2017-09-28 CRAN (R 3.4.2)                    
#>  evaluate     0.10.1     2017-06-24 CRAN (R 3.4.1)                    
#>  forcats    * 0.3.0      2018-02-19 cran (@0.3.0)                     
#>  foreign      0.8-69     2017-06-22 CRAN (R 3.4.4)                    
#>  ggplot2    * 2.2.1.9000 2018-03-21 Github (tidyverse/ggplot2@39e4a3b)
#>  glue         1.2.0      2017-10-29 CRAN (R 3.4.2)                    
#>  graphics   * 3.4.4      2018-03-15 local                             
#>  grDevices  * 3.4.4      2018-03-15 local                             
#>  grid         3.4.4      2018-03-15 local                             
#>  gtable       0.2.0      2016-02-26 CRAN (R 3.4.0)                    
#>  haven        1.1.1      2018-01-18 CRAN (R 3.4.3)                    
#>  here       * 0.1        2017-05-28 CRAN (R 3.4.0)                    
#>  hms          0.4.2      2018-04-09 Github (tidyverse/hms@c0cfc01)    
#>  htmltools    0.3.6      2017-04-28 CRAN (R 3.4.0)                    
#>  httr         1.3.1      2017-08-20 CRAN (R 3.4.1)                    
#>  jsonlite     1.5        2017-06-01 CRAN (R 3.4.0)                    
#>  knitr        1.20       2018-02-20 CRAN (R 3.4.3)                    
#>  lattice      0.20-35    2017-03-25 CRAN (R 3.4.4)                    
#>  lazyeval     0.2.1      2017-10-29 CRAN (R 3.4.2)                    
#>  lubridate  * 1.7.3      2018-02-27 CRAN (R 3.4.3)                    
#>  magrittr     1.5        2014-11-22 CRAN (R 3.4.0)                    
#>  memoise      1.1.0      2017-04-21 CRAN (R 3.4.0)                    
#>  methods    * 3.4.4      2018-03-15 local                             
#>  mnormt       1.5-5      2016-10-15 CRAN (R 3.4.0)                    
#>  modelr       0.1.1      2017-07-24 CRAN (R 3.4.1)                    
#>  munsell      0.4.3      2016-02-13 CRAN (R 3.4.0)                    
#>  nlme         3.1-131.1  2018-02-16 CRAN (R 3.4.4)                    
#>  parallel     3.4.4      2018-03-15 local                             
#>  pillar       1.2.1      2018-02-27 CRAN (R 3.4.3)                    
#>  pkgconfig    2.0.1      2017-03-21 CRAN (R 3.4.0)                    
#>  plyr         1.8.4      2016-06-08 CRAN (R 3.4.0)                    
#>  psych        1.7.8      2017-09-09 CRAN (R 3.4.2)                    
#>  purrr      * 0.2.4      2017-10-18 CRAN (R 3.4.2)                    
#>  R6           2.2.2      2017-06-17 CRAN (R 3.4.0)                    
#>  Rcpp         0.12.16    2018-03-13 CRAN (R 3.4.3)                    
#>  readr      * 1.2.0      2018-04-09 Github (tidyverse/readr@23eb46d)  
#>  readxl       1.0.0      2017-04-18 CRAN (R 3.4.0)                    
#>  reshape2     1.4.3      2017-12-11 cran (@1.4.3)                     
#>  rlang        0.2.0.9001 2018-04-16 Github (tidyverse/rlang@82b2727)  
#>  rmarkdown    1.9.8      2018-04-20 Github (rstudio/rmarkdown@81dabbc)
#>  rprojroot    1.3-2      2018-01-03 CRAN (R 3.4.3)                    
#>  rvest        0.3.2      2016-06-17 CRAN (R 3.4.0)                    
#>  scales       0.5.0.9000 2018-03-21 Github (hadley/scales@d767915)    
#>  stats      * 3.4.4      2018-03-15 local                             
#>  stringi      1.1.7      2018-03-12 CRAN (R 3.4.4)                    
#>  stringr    * 1.3.0      2018-02-19 cran (@1.3.0)                     
#>  tibble     * 1.4.2      2018-01-22 cran (@1.4.2)                     
#>  tidyr      * 0.8.0      2018-01-29 cran (@0.8.0)                     
#>  tidyverse  * 1.2.1      2017-11-14 CRAN (R 3.4.2)                    
#>  tools        3.4.4      2018-03-15 local                             
#>  utils      * 3.4.4      2018-03-15 local                             
#>  withr        2.1.2      2018-03-21 Github (jimhester/withr@79d7b0d)  
#>  xml2         1.2.0      2018-01-24 CRAN (R 3.4.3)                    
#>  yaml         2.1.18     2018-03-08 CRAN (R 3.4.4)
```

</details>

**Comment by yihui** [comment]
Unfortunately, the short answer is no: the default site generator in **rmarkdown** uses the same R session and environment for all Rmd files. Yes, it could cause problems, and it is kind of by design.

blogdown compiles each Rmd file in a different R session; bookdown has two ways to compile Rmd files: https://bookdown.org/yihui/bookdown/new-session.html

The only possible way to go is a custom site generator (https://rmarkdown.rstudio.com/rmarkdown_site_generators.html), in which you call `Rscript` to render each Rmd in a new R session. I don't know how difficult it is to implement this.

**Comment by andrewheiss** [comment]
Cool cool cool. I'll look into site generators and see what I can figure out. Thanks!

**Comment by rich-iannone** [comment]
Since some good info was passed around—and there's nothing left to do here—I'm going to close this issue.

**Comment by jdblischak** [comment]
In case anyone else is interested, I put together a quick prototype of a custom site generator that uses `callr::r_safe` to isolate the execution of each R Markdown file of a website created with `rmarkdown::render_site`. See https://github.com/andrewheiss/rmarkdown-website-envs/pull/1

I also note that an alternative to callr would be to use the function [Rscript](https://github.com/yihui/xfun/blob/0283f65c9fe34f4e15ae6c333b470d81656245ff/R/command.R#L13) @yihui's xfun package (with the caveat that this function is currently only available in the development version of xfun on GitHub).

**Comment by andrewheiss** [comment]
Amazing, thanks!

I've merged in the PR, which now makes that repo a package. For historical reasons (and to make it so people don't have to go back through git history), I've made a copy of the original reprex repository at https://github.com/andrewheiss/rmarkdown-website-envs-orig

**Comment by yihui** [comment]
FYI this has been implemented in #1443.

**Comment by andrewheiss** [comment]
Oh awesome, this is great!
